File: Two Pointer/boj_2003_수들의 합2.cpp
//https://www.acmicpc.net/problem/2003
#include <iostream>

using namespace std;

int arr[10000];

int main() {
	ios::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);
	int n, m;
	int answer = 0;
	cin >> n >> m;
	for (int i = 0; i < n; i++) {
		cin >> arr[i];
	}
	int p1 = 0, p2 = 0;
	int sum = 0;
	while (p1 < n) {
		if (sum == m) {
			answer++;
		}
		else if (sum > m) {
			sum-=arr[p1++];
			continue;
		}

		if (p2 < n) {
			sum += arr[p2++];
		}
		else {
			sum -= arr[p1++];
		}

	}
	cout << answer;
}

File: Two Pointer/boj_2003_수들의 합2.kt
//https://www.acmicpc.net/problem/2003
import java.util.StringTokenizer

fun main() = with(System.out.bufferedWriter()){
    val br = System.`in`.bufferedReader()
    var tk = StringTokenizer(br.readLine())
    val (n,m) = List(2){ Integer.parseInt(tk.nextToken())};
    tk = StringTokenizer(br.readLine())
    val arr = IntArray(n)
    for(i in 0 until n){
        arr[i]= Integer.parseInt(tk.nextToken())
    }
    var s=0
    var e=0
    var sum =0
    var answer=0

    while(s<n){
        if(sum>m){
            sum -=arr[s++]
        }
        else if(sum==m){
            sum-=arr[s++]
            answer++
        }
        else{
            if(e <n){
                sum += arr[e++]
            }
            else {
                sum -= arr[s++]
            }
        }
    }
    write("$answer")
    close()
}
